Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+    
+    mutating func receive(metadata: Metadata, endStream: Bool) throws -> OnMetadataReceived {
+      // This is metadata received by the client from the server.
+      // It can be initial, which confirms that the server is now open;
+      // or an END_STREAM trailer, meaning the response is over.
+      if endStream {
+        try self.clientReceivedEndHeader()
+      } else {
+        try self.clientReceivedMetadata()
+      }
+      return .doNothing
+    }
+    
+    mutating func clientReceivedEndHeader() throws {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't have received a stream end trailer if both client and server are idle.")
+      case .clientOpenServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent an end stream header if it is still idle.")
+      case .clientOpenServerOpen(let state):
+        self.state = .clientOpenServerClosed(.init(previousState: state))
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is already closed, can't have received the end stream trailer twice.")
+      case .clientClosedServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent end stream trailer if it is idle.")
+      case .clientClosedServerOpen(let state):
+        state.compressor?.end()
+        state.decompressor?.end()
+        self.state = .clientClosedServerClosed
+      case .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent end stream trailer if it is already closed.")
+      }
+    }
+    
+    mutating func clientReceivedMetadata() throws {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent metadata if the client is idle.")
+      case .clientOpenServerIdle(let state):
+        self.state = .clientOpenServerOpen(.init(previousState: state))
+      case .clientOpenServerOpen:
+        // This state is valid: server can send trailing metadata without END_STREAM
+        // set, and follow it with an empty message frame where the flag *is* set.
+        ()
+        // TODO: I believe we should set some flag in the state to signal that
+        // we're expecting an empty data frame with END_STREAM set; otherwise,
+        // we could get an infinite number of metadata frames from the server -
+        // not sure this should be allowed.
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is closed, nothing could have been sent.")
+      case .clientClosedServerClosed, .clientClosedServerIdle, .clientClosedServerOpen:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is closed, shouldn't have received anything.")
```

Comment:
This isn't true: if the client is closed it means it's sent end stream, it can still receive messages until the server sends end stream. `.clientClosedServerIdle` is a valid state here and `.clientClosedServerOpen` needs to be treated like `.clientOpenServerOpen`.